### PR TITLE
osd: ECBackend::get_hash_info() makes use of the obc::attr_cache

### PR DESCRIPTION
**Pull request by rzarzynski**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/common/scrub_types.cc [hunk 1]
```diff
@@ -55,11 +55,8 @@ static void encode(const osd_shard_t& shard, bufferlist& bl) {
 
 void shard_info_wrapper::set_object(const ScrubMap::object& object)
 {
-  for (auto attr : object.attrs) {
-    bufferlist bl;
-    bl.push_back(attr.second);
-    attrs.insert(std::make_pair(attr.first, std::move(bl)));
-  }
+  // no-op, changes the comparator from std::less<void>
+  attrs = reinterpret_cast<const std::map<std::string, ceph::bufferlist>&>(object.attrs);
```
Comment (rzarzynski) [inline src/common/scrub_types.cc:59]: I dislike it.

Comment (mattbenjamin) [reply src/common/scrub_types.cc:59]: that is naughty;  it can't be a static_cast?

Comment (tchaikov) [reply src/common/scrub_types.cc:59]: the type of `ScrubMap::object::attrs` is `std::map<std::string, ceph::buffer::ptr, std::less<>>`, while the type of `shard_info_t::attrs` is `std::map<std::string, ceph::bufferlist>`.

why do you think `ceph::buffer::ptr` can be used as `ceph::bufferlist` ? please note that the latter is an alias of `ceph::buffer::list`

Comment (rzarzynski) [reply src/common/scrub_types.cc:59]: Unfortunately, `static_cast` doesn't compile. I will take a look on the transparent comparator and elaborate on changes. Thanks, Guys!

Comment (rzarzynski) [reply src/common/scrub_types.cc:59]: My understanding is `shard_info_t` is a part of the public API but rather infrequently used.

Its sole client is `get_inconsistent_objects()`. The struct is used indirectly, through the `std::vector<inconsistent_obj_t>* objects` parameter.

```cpp
    /**
     * List the inconsistent objects found in a given PG by last scrub
     *
     * @param pg the placement group returned by @c pg_list()
     * @param start_after the first returned @c objects
     * @param max_return the max number of the returned @c objects
     * @param c what to do when the operation is complete and safe
     * @param objects [out] the objects where inconsistencies are found
     * @param interval [in,out] an epoch indicating current interval
     * @returns if a non-zero @c interval is specified, will return -EAGAIN i
     *          the current interval begin epoch is different.
     */
    int get_inconsistent_objects(const PlacementGroup& pg,
                                 const object_id_t &start_after,
                                 unsigned max_return,
                                 AioCompletion *c,
                                 std::vector<inconsistent_obj_t>* objects,
                                 uint32_t* interval);
```

```cpp
struct inconsistent_obj_t : obj_err_t {
...
  std::map<osd_shard_t, shard_info_t> shards;
...
};
```

If we enable the transparent comparator `shard_info_t::attrs` there is non-zero risk breaking buildability of 3rd party applications, though the risk is very small (negligible?).

Comment (athanatos) [reply src/common/scrub_types.cc:59]: Can we use a loop to fill attrs somewhat inefficiently and worry about this later?

File: src/osd/ECBackend.cc [hunk 2]
```diff
@@ -1103,7 +1103,14 @@ void ECBackend::handle_sub_read(
 	// Do NOT check osd_read_eio_on_bad_digest here.  We need to report
 	// the state of our chunk in case other chunks could substitute.
         ECUtil::HashInfoRef hinfo;
-        hinfo = get_hash_info(i->first);
+        map<string,bufferlist,less<>> attrs;
```
Comment (athanatos) [inline src/osd/ECBackend.cc:1106]: spaces after ,

**Review by athanatos — APPROVED** [review]
LGTM other than formatting nits and shard_info_wrapper::set_object

File: src/common/scrub_types.cc (hunk 1 above)
Comment (rzarzynski) [reply src/common/scrub_types.cc:59]: Yeah, we should start from the loop; `reinterpret_cast` looks like a premature optimization.

**Comment by rzarzynski** [comment]
jenkins retest this please

File: src/common/scrub_types.cc (hunk 1 above)
Comment (rzarzynski) [reply src/common/scrub_types.cc]: Switched to the loop (and left a comment).

File: src/osd/ECBackend.cc (hunk 2 above)
Comment (rzarzynski) [reply src/osd/ECBackend.cc]: Fixed!

File: src/osd/ECBackend.cc [hunk 3]
```diff
@@ -1912,49 +1920,42 @@ void ECBackend::do_read_op(ReadOp &op)
 }
 
 ECUtil::HashInfoRef ECBackend::get_hash_info(
-  const hobject_t &hoid, bool create, const map<string,bufferlist,less<>>& attrs)
+  const hobject_t &hoid,
+  bool create,
+  const map<string,bufferlist,less<>>& attrs,
```
Comment (rzarzynski) [inline src/osd/ECBackend.cc:1925]: nit.

File: src/osd/ECBackend.cc [hunk 4]
```diff
@@ -1104,9 +1104,13 @@ void ECBackend::handle_sub_read(
...
+	struct stat st;
         int r = PGBackend::objects_get_attrs(i->first, &attrs);
         if (r >= 0) {
-          hinfo = get_hash_info(i->first, false, attrs);
+	  r = object_stat(i->first, &st);
+	}
+	if (r >= 0) {
+          hinfo = get_hash_info(i->first, false, attrs, st.st_size);
 	} else {
           derr << "get_hash_info" << ": stat " << i->first << " failed: "
```
Comment (rzarzynski) [inline src/osd/ECBackend.cc:1115]: err handling.

**Comment by rzarzynski** [comment]
```
1 tests failed out of 288

Total Test time (real) = 2926.95 sec

The following tests FAILED:
	  7 - run-tox-mgr-dashboard-py3 (Failed)
```

**Comment by rzarzynski** [comment]
jenkins test make check

**Pull request closed without merging**